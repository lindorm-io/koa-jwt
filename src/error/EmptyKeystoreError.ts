import { APIError } from "@lindorm-io/errors";
import { HttpStatus } from "@lindorm-io/core";

export class EmptyKeystoreError extends APIError {
  public constructor() {
    super("Keystore was initialised without keys", {
      statusCode: HttpStatus.ServerError.INTERNAL_SERVER_ERROR,
    });
  }
}
