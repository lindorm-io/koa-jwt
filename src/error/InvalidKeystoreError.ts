import { APIError } from "@lindorm-io/errors";
import { HttpStatus } from "@lindorm-io/core";

export class InvalidKeystoreError extends APIError {
  public constructor() {
    super("Keystore could not be found on context", {
      statusCode: HttpStatus.ServerError.INTERNAL_SERVER_ERROR,
    });
  }
}
