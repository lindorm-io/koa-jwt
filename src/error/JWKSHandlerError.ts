import { ExtendableError } from "@lindorm-io/global";

export class JWKSHandlerError extends ExtendableError {
  constructor(error: Error, url: string) {
    super("Unable to find JWKS on URL", {
      debug: { error, url },
    });
  }
}
