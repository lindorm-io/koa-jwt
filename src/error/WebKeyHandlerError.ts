import { ExtendableError } from "@lindorm-io/core";

export class WebKeyHandlerError extends ExtendableError {
  constructor(error: Error, url: string) {
    super("Unable to find JWKS on URL", {
      debug: { error, url },
    });
  }
}
