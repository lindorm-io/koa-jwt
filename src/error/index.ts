export * from "./EmptyKeystoreError";
export * from "./InvalidKeystoreError";
