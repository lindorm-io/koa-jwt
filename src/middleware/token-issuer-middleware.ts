import { EmptyKeystoreError, InvalidKeystoreError } from "../error";
import { Middleware } from "@lindorm-io/koa";
import { TokenIssuer } from "@lindorm-io/jwt";
import { TokenIssuerContext } from "../types";

interface Options {
  issuer: string;
  issuerName: string;
  keystoreName: string;
}

export const tokenIssuerMiddleware =
  (options: Options): Middleware<TokenIssuerContext> =>
  async (ctx, next): Promise<void> => {
    const start = Date.now();

    if (!ctx.keystore || !ctx.keystore[options.keystoreName]) {
      throw new InvalidKeystoreError();
    }
    if (ctx.keystore[options.keystoreName].getUsableKeys().length === 0) {
      throw new EmptyKeystoreError();
    }

    ctx.issuer = {
      ...(ctx.issuer || {}),
      [options.issuerName]: new TokenIssuer({
        issuer: options.issuer,
        keystore: ctx.keystore[options.keystoreName],
        logger: ctx.logger,
      }),
    };

    ctx.metrics.tokenIssuer = Date.now() - start;

    await next();
  };
