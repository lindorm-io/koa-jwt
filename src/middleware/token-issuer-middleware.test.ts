import MockDate from "mockdate";
import { EmptyKeystoreError, InvalidKeystoreError } from "../error";
import { Keystore } from "@lindorm-io/key-pair";
import { getTestKeystore, logger } from "../test";
import { tokenIssuerMiddleware } from "./token-issuer-middleware";

MockDate.set("2020-01-01T08:00:00.000Z");

const next = jest.fn();

describe("tokenIssuerMiddleware", () => {
  let ctx: any;
  let options: any;

  beforeEach(() => {
    ctx = {
      keystore: {
        keystoreName: getTestKeystore(),
      },
      metrics: {},
      logger,
    };
    options = {
      issuer: "issuer",
      issuerName: "issuerName",
      keystoreName: "keystoreName",
    };
  });

  test("should set issuer on context", async () => {
    await expect(tokenIssuerMiddleware(options)(ctx, next)).resolves.toBeUndefined();
    expect(ctx.issuer).toMatchSnapshot();
  });

  test("should throw InvalidKeystoreError", async () => {
    ctx = { logger };
    await expect(tokenIssuerMiddleware(options)(ctx, next)).rejects.toStrictEqual(expect.any(InvalidKeystoreError));
  });

  test("should throw InvalidKeystoreError", async () => {
    ctx = { keystore: { keystoreName: new Keystore({ keys: [] }), logger } };
    await expect(tokenIssuerMiddleware(options)(ctx, next)).rejects.toStrictEqual(expect.any(EmptyKeystoreError));
  });
});
