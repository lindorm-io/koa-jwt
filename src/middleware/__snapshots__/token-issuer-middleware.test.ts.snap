// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenIssuerMiddleware should set issuer on context 1`] = `
Object {
  "issuerName": TokenIssuer {
    "issuer": "issuer",
    "keystore": Keystore {
      "keys": Array [
        Object {
          "algorithms": Array [
            "ES512",
          ],
          "allowed": true,
          "created": 2020-01-01T08:00:00.000Z,
          "events": Array [],
          "expires": null,
          "id": "7531da89-12e9-403e-925a-5da49100635c",
          "namedCurve": "P-521",
          "passphrase": null,
          "preferredAlgorithm": "ES512",
          "privateKey": "-----BEGIN PRIVATE KEY-----
MIHuAgEAMBAGByqGSM49AgEGBSuBBAAjBIHWMIHTAgEBBEIBGma7xGZpaAngFXf3
mJF3IxZfDpI+6wU564K+eehxX104v6dZetjSfMx0rvsYX/s6cO2P3GE7R95VxWEk
+f4EX0qhgYkDgYYABAB8cBfDwCi41G4kVW4V3Y86nIMMCypYzfO8gYjpS091lxkM
goTRS3LM1p65KQfwBolrWIdVrbbOILASf06fQsHw5gEt4snVuMBO+LS6pesX9vA8
QT1LjX75Xq2InnLY1VToeNmxkuM+oDZgqHOYwzfUhu+zZaA5AuEkqPi47TA9iCSY
VQ==
-----END PRIVATE KEY-----
",
          "publicKey": "-----BEGIN PUBLIC KEY-----
MIGbMBAGByqGSM49AgEGBSuBBAAjA4GGAAQAfHAXw8AouNRuJFVuFd2POpyDDAsq
WM3zvIGI6UtPdZcZDIKE0UtyzNaeuSkH8AaJa1iHVa22ziCwEn9On0LB8OYBLeLJ
1bjATvi0uqXrF/bwPEE9S41++V6tiJ5y2NVU6HjZsZLjPqA2YKhzmMM31Ibvs2Wg
OQLhJKj4uO0wPYgkmFU=
-----END PUBLIC KEY-----
",
          "type": "EC",
          "updated": 2020-01-01T08:00:00.000Z,
          "version": 0,
        },
      ],
    },
    "logger": Object {
      "createChildLogger": [Function],
      "debug": [Function],
      "error": [Function],
      "info": [Function],
      "silly": [Function],
      "verbose": [Function],
      "warn": [Function],
    },
  },
}
`;
